### PR TITLE
Post picker: Design updates, manage current selection, re-order selection. 

### PR DESCRIPTION
**Pull request by mattheu**
First round of changes here are just functional, focused on getting everything working. 

This round of changes is a bit more opinionated, hence why I'm opening in a separate pull request. 

* Add tab UI to switch between browsing posts and viewing the current selection
* Ability to reorder selection from the current selection tab. 
* Break things up into separate components.
* New ToolbarButton component. 
* Some design adjustments. 

Some screenshots: 

image

image

### CODE REVIEW

File: src/components/TermSelector/index.js [hunk 1]
```diff
@@ -47,26 +47,27 @@ function generateIdToTitleMap( termObjects ) {
...
 			taxonomyTermsByTitle,
 		};
-
 	}, [ taxonomy ] );
 
 	const selectedTerms = value.map( id => taxonomyTermsById[id] );
 
 	return ( <FormTokenField
-		label={ `Filter by ${taxonomy}` }
+		label={ `Filter by ${ taxObject ? taxObject.labels.singular_name : '' }` }
```
Comment (tomjn) [inline src/components/TermSelector/index.js:70]: this should be translatable with a placeholder

File: src/components/PostPicker/index.js [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+			align="flex-start" s
+			style={ { gap: 24 } }
+		>
+			<FlexItem
+				style={ {
+					width: '35%',
+				} }
+			>
+				<SearchControl
+					label="Search Posts"
```
Comment (tomjn) [inline src/components/PostPicker/index.js:217]: this should be translatable

File: src/components/PostPicker/index.js [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+
+	const queryArgs = {
+		search,
+		per_page: 30,
+		...taxQueries,
+	};
+
+	return (
+		<Flex
+			align="flex-start" s
```
Comment (tomjn) [inline src/components/PostPicker/index.js:208]: stray `s`
```suggestion
			align="flex-start"
```

File: src/components/PostPicker/index.js [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+	 * Sets correct property value of rest_base.
+	 * Merges the new data with existing data.
+	 *
+	 * @param {*} taxonomy Taxonomy.
+	 * @param {*} newTerms New terms.
+	 */
+	const updateTaxQueryState = useCallback( ( taxonomy, newTerms ) => {
+		const taxObject = taxObjects.find( t => t && t.slug === taxonomy );
+
+		if ( taxObject ) {
```
Comment (tomjn) [inline src/components/PostPicker/index.js:182]: would this be easier to read if it returned early? I know it's an extra line of code but it'd reduce indenting

File: src/components/PostPicker/index.js [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+ */
+function BrowsePanel( props ) {
+	const {
+		postType,
+		onChange,
+		values,
+		taxonomies,
+	} = props;
+
+	const [ search, setSearch ] = useState();
```
Comment (tomjn) [inline src/components/PostPicker/index.js:162]: does this need an initial value?

**Review by tomjn — CHANGES_REQUESTED** [review]
Some changes re: localisation, there's eslint issues too

File: src/components/PostPicker/index.js (hunk 5 above)
Comment (mattheu) [reply src/components/PostPicker/index.js:162]: Hmm... maybe. 

If `search` is an empty string, or `null`, and that is passed as part of the `queryArgs` to `getEntityRecords` then the API request includes `?search=`. I don't _think_ that's a problem... I assume the Rest API is ignoring an empty search query. However it does feel _neater_ if we dont' send this arg when it's empty. 

But -  you can also get issues if values are undefined. Passing them straight to an `input` as the value, will cause a warning like `Warning: A component is changing an uncontrolled input of type text to be controlled.`. 

So probably it is better to be consistent about the type of data we're dealing with, setting an empty string as a default, and then handle that when defining query args.

**Comment by mattheu** [comment]
Thanks for your feedback @tomjn , I think I addressed everything.

File: .eslintignore [hunk 6]
```diff
@@ -1,3 +1,4 @@
 **/node_modules/
 /.github/
 /dist/
+e
```
Comment (Sephsekla) [inline .eslintignore:4]: Is this deliberate?

File: src/components/PostPicker/index.js [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+	FlexItem,
+	Modal,
+	Notice,
+	SearchControl,
+	Spinner,
+} from '@wordpress/components';
+import { useSelect } from '@wordpress/data';
+import {
+	useState,
+	useMemo,
```
Comment (Sephsekla) [inline src/components/PostPicker/index.js:19]: As flagged by the linter, no need for useMemo here.

File: src/components/PostPicker/index.js [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+	const {
+		postType,
+		queryArgs,
+		onChange,
+		values = [],
+		isSortable = false,
+	} = props;
+
+	const queriedPosts = useSelect( ( select ) => {
+		return select( 'core' ).getEntityRecords( 'postType', postType, queryArgs ) ?? [];
```
Comment (Sephsekla) [inline src/components/PostPicker/index.js:41]: On WP 6.1, we could make use of useEntityRecords (although the project I originally wrote this component for is still on 6.0). But it raises a question, which is should we clarify our supported WP versions for this library? I feel like it's eventually going to be an issue if we're not targeting a clear range.

File: src/components/PostPicker/index.js [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+					)
+			}
+		</div>
+	);
+}
+
+/**
+ * Post Picker Browse Post Panel.
+ *
+ * @param props
```
Comment (Sephsekla) [inline src/components/PostPicker/index.js:151]: Missing type and description for `props`

**Review by Sephsekla — COMMENTED** [review]
This is really nice, some decent improvements here.

I've only given it a quick look over, but spotted there are a few things being flagged by eslint. 

Does this also contain the commits from https://github.com/humanmade/block-editor-components/pull/142? I'm wondering if it makes sense to make this PR against that branch just to make it clearer what's being changed here vs the other pull request.

File: src/components/PostPicker/index.js (hunk 8 above)
Comment (mattheu) [reply src/components/PostPicker/index.js:41]: Good thinking. We should agree on this separate to this PR and document it. We could align with Altis support of about a year from Altis release date which would mean 6.0 is supported (Until October) and probably should leave this as-is.

**Comment by mattheu** [comment]
@Sephsekla Feedback addressed. 

> Does this also contain the commits from https://github.com/humanmade/block-editor-components/pull/142? I'm wondering if it makes sense to make this PR against that branch just to make it clearer what's being changed here vs the other pull request.

Yep. Actually as #142 was approved, I merged that into `post-picker`. But same outcome 👍

**Comment by tomjn** [comment]
@mattheu @Sephsekla is this PR still active or were its changes merged in? Would be good to get this merged/closed

**Comment by mattheu** [comment]
Yep good call. Will merge this to #141 now.

**Pull request merged**